utils/pathfinder-probe/src/main.rs

```diff
@@ -0,0 +1,112 @@
+            }
+            tokio::time::sleep(setup.poll_delay).await;
+        }
+    });
+
+    let app = Router::new().route("/metrics", routing::get(|| async move { handle.render() }));
+
+    axum::Server::bind(&listen_at)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct Setup {
+    listen_at: String,
+    gateway_url: String,
+    pathfinder_url: String,
+    poll_delay: Duration,
+}
+
+fn setup() -> anyhow::Result<Setup> {
+    let args = std::env::args().collect::<Vec<String>>();
+    args.get(1)
+        .zip(args.get(2))
+        .zip(args.get(3))
+        .zip(args.get(4))
+        .map(|(((listen_at, gateway_url), pathfinder_url), delay_seconds)| Ok(Setup {
+            listen_at: listen_at.to_string(),
+            gateway_url: gateway_url.to_string(),
+            pathfinder_url: pathfinder_url.to_string(),
+            poll_delay: Duration::from_secs(delay_seconds.parse().context("Failed to parse <poll-seconds> integer")?),
+        }))
+        .ok_or(anyhow::anyhow!("Failed to parse arguments: <listen-at> <gateway-url> <pathfinder-url> <poll-delay-seconds>"))?
+}
+
+// curl "https://alpha-mainnet.starknet.io/feeder_gateway/get_block?blockNumber=latest" 2>/dev/null | jq '.block_number'
+async fn get_gateway_latest(gateway_url: &str) -> anyhow::Result<i64> {
+    let json: serde_json::Value = reqwest::ClientBuilder::new()
+        .build()?
+        .get(&format!(
+            "{}/feeder_gateway/get_block?blockNumber=latest",
+            gateway_url
+        ))
+        .send()
+        .await?
+        .json()
+        .await?;
+
+    json["block_number"]
+        .as_i64()
+        .ok_or(anyhow::anyhow!("Failed to fetch block number"))
+}
+
+// curl -H 'Content-type: application/json' -d '{"jsonrpc":"2.0","method":"starknet_blockNumber","params":[],"id":1}' http://127.0.0.1:9000/rpc/v0.3
+async fn get_pathfinder_head(pathfinder_url: &str) -> anyhow::Result<i64> {
+    let json: serde_json::Value = reqwest::ClientBuilder::new().build()?
+        .post(&format!("{}/rpc/v0.3", pathfinder_url))
+        .header("Content-type", "application/json")
+        .json(&serde_json::json!({"jsonrpc":"2.0","method":"starknet_blockNumber","params":[],"id":1}))
+        .send()
+        .await?
+        .json()
+        .await?;
+
+    json["result"]
+        .as_i64()
+        .ok_or(anyhow::anyhow!("Failed to fetch block number"))
+}
+
+async fn tick(setup: &Setup) -> anyhow::Result<()> {
+    let gw_head = get_gateway_latest(&setup.gateway_url).await?;
+    tracing::debug!(head = gw_head, "gateway");
+
+    let pf_head = get_pathfinder_head(&setup.pathfinder_url).await?;
+    tracing::debug!(head = pf_head, "pathfinder");
+
+    metrics::gauge!("gw_head", gw_head as f64);
+    metrics::gauge!("pf_head", pf_head as f64);
+
+    let delay = gw_head - pf_head;
+    metrics::gauge!("blocks_missing", delay as f64);
+
+    Ok(())
+}
```

Comment:
This won't work because it will block on either the pathfinder or gateway queries? So e.g. if pathfinder is down you just get no more information back. Or are you relying on the metrics consumer to detect that case?

It also only informs us of the difference in block count, not in time difference between the two.

Reply:
Time difference is already being reported (difference between the most recent and previous head blocks) from Pathfinder directly.

Reply:
The dashboard: http://34.175.234.24:3000/d/c45061a3-8291-44f8-8dcf-e47a166ac798/pathfinder?orgId=1&refresh=30s&from=now-24h&to=now

<img width="698" alt="Screenshot 2023-07-25 at 23 18 22" src="https://github.com/eqlabs/pathfinder/assets/8093171/88bf62f7-dfd8-4630-badc-b49777c6b10b">

Reply:
> Time difference is already being reported (difference between the most recent and previous head blocks) from Pathfinder directly.

That only works while pathfinder itself is still running. What I would imagine is the probe should actually be reporting this -- and not really using any pathfinder metrics at all.

What happens for example if block times are extremely slow, but pathfinder has stopped syncing. Would it not happen that pathfinder's latest block reports a reasonably low latency metric. But in reality we are a block or two behind, which are already many minutes old? But we wouldn't be able to tell this from the current metrics at all? Apart from being 1 or 2 blocks behind which would seem perfectly reasonable.

What I was thinking is storing the last `N` block height, hash and query time instance for both gateway and pathfinder -- which would let you assemble any metric / time delay etc locally in the probe.

Reply:
Makes sense. Added a `blocks_delay` metrics based only on probe's data.